src/latbin/lattice-lmrescore-kaldi-rnnlm-pruned-cache-adapt-onlyhistory.cc

```diff
@@ -0,0 +1,414 @@
+  void OutputPosteriors(
+                        std::map<int, double> *word_to_count,
+                        std::map<int, double> *word_to_count2,
+                        double* sum,
+                        double* sum2) {
+    int32 num_post = 0;
+    if (!ComputeCompactLatticeAlphas(clat_, &alpha_))
+      return;
+    if (!ComputeCompactLatticeBetas(clat_, &beta_))
+      return;
+
+    CompactLatticeStateTimes(clat_, &state_times_);
+    if (clat_.Start() < 0)
+      return;
+    double tot_like = beta_[clat_.Start()];
+
+    int32 num_states = clat_.NumStates();
+    for (int32 state = 0; state < num_states; state++) {
+      for (fst::ArcIterator<CompactLattice> aiter(clat_, state);
+           !aiter.Done(); aiter.Next()) {
+        const CompactLatticeArc &arc = aiter.Value();
+        double arc_loglike = -ConvertToCost(arc.weight) +
+            alpha_[state] + beta_[arc.nextstate] - tot_like;
+        KALDI_ASSERT(arc_loglike < 0.1 &&
+                     "Bad arc posterior in forward-backward computation");
+        if (arc_loglike > 0.0) arc_loglike = 0.0;
+
+        int32 word = arc.ilabel;
+        BaseFloat arc_post = exp(arc_loglike);
+        if (arc_post <= min_post_) continue;
+
+        (*word_to_count)[word] += arc_post;
+        (*word_to_count2)[word] += arc_post;
+        *sum += arc_post;
+        *sum2 += arc_post;
+        num_post++;
+      }
+    }
+  }
+ private:
+  const CompactLattice &clat_;
+  std::vector<double> alpha_;
+  std::vector<double> beta_;
+  std::vector<int32> state_times_;
+
+  BaseFloat min_post_;
+};
+}  // namespace kaldi
+
+void ReadUttToConvo(string filename, map<string, string> &m) {
+  KALDI_ASSERT(m.size() == 0);
+  ifstream ifile(filename.c_str());
+  string utt, convo;
+  while (ifile >> utt >> convo) {
+    m[utt] = convo;
+  }
+}
+
+void ReadUnigram(string filename, std::vector<double> *unigram) {
+  std::vector<double> &m = *unigram;
+  ifstream ifile(filename.c_str());
+  int32 word;
+  double count;
+  double sum = 0.0;
+  while (ifile >> word >> count) {
+    m[word] = count;
+    sum += count;
+  }
+
+  for (int32 i = 0; i < m.size(); i++) {
+    m[i] /= sum;
+  }
+}
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+    using fst::SymbolTable;
+    using fst::VectorFst;
+    using fst::StdArc;
+    using fst::ReadFstKaldi;
+
+    const char *usage =
+        "Rescores lattice with kaldi-rnnlm. This script is called from \n"
+        "scripts/rnnlm/lmrescore_pruned.sh. An example for rescoring \n"
+        "lattices is at egs/swbd/s5c/local/rnnlm/run_lstm.sh \n"
+        "\n"
+        "Usage: lattice-lmrescore-kaldi-rnnlm-pruned-cache-adapt-onlyhistory [options] \\\n"
+        "             <old-lm-rxfilename> <embedding-file> \\\n"
+        "             <raw-rnnlm-rxfilename> \\\n"
+        "             <lattice-rspecifier> <lattice-wspecifier>\n"
```

Comment:
I notice the usage message of this program does not match the actual usage.  Could you please clean it up a bit?

Reply:
@danpovey I have modified the usage part and refactored the code. 

Ke